src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -66,60 +109,244 @@ protected AbstractEntityAIStructure(@NotNull final J job)
+                }
+                resetTask();
+            }
+            else if(job instanceof JobMiner)
+            {
+                BuildingMiner minerBuilding = (BuildingMiner) getOwnBuilding();
+                //If shaft isn't cleared we're in shaft clearing mode.
+                if(!minerBuilding.clearedShaft)
+                {
+                    @NotNull final Level currentLevel = new Level(minerBuilding, ((JobMiner) job).getStructure().getPosition().getY());
+                    minerBuilding.addLevel(currentLevel);
+                    minerBuilding.setCurrentLevel(minerBuilding.getNumberOfLevels());
+                    minerBuilding.resetStartingLevelShaft();
+
+                    //Send out update to client
+                    getOwnBuilding().markDirty();
+                }
+                else
+                {
+                    minerBuilding.getCurrentLevel().closeNextNode(rotation);
+                }
+                ((JobMiner) job).setStructure(null);
+            }
+            worker.addExperience(XP_EACH_BUILDING);
+            workFrom = null;
+        }
+
+        currentStructure = null;
+
+        return AIState.IDLE;
+    }
+
+    private Boolean decorationStep(final Structure.StructureBlock structureBlock)
+    {
+        if (!BlockUtils.shouldNeverBeMessedWith(structureBlock.worldBlock))
+        {
+            //Fill workFrom with the position from where the builder should build.
+            //also ensure we are at that position.
+            if (!walkToConstructionSite())
+            {
+                return false;
+            }
+
+            if (structureBlock.block == null
+                    || structureBlock.doesStructureBlockEqualWorldBlock()
+                    || structureBlock.metadata.getMaterial().isSolid())
+            {
+                //findNextBlock count was reached and we can ignore this block
+                return true;
+            }
+
+            worker.faceBlock(structureBlock.blockPosition);
+
+            @Nullable final Block block = structureBlock.block;
+
+            //should never happen
+            if (block == null)
+            {
+                @NotNull final BlockPos local = structureBlock.blockPosition;
+                Log.getLogger().error(String.format("StructureProxy has null block at %s - local(%s)", currentStructure.getCurrentBlockPosition(), local));
+                return true;
+            }
+
+            @Nullable final IBlockState blockState = structureBlock.metadata;
+            //We need to deal with materials
+            if (!Configurations.builderInfiniteResources
+                    && !handleMaterials(block, blockState))
+            {
+                return false;
+            }
+
+            placeBlockAt(block, blockState, structureBlock.blockPosition);
+        }
+        return true;
+    }
+
+    private Boolean structureStep(final Structure.StructureBlock structureBlock)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "structureStep" is 19 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)